### PR TITLE
Persistent concurrent hash map data structure.

### PR DESCRIPTION
**Pull request by vinser52**
Implementation is based on concurrent_hash_map from Intel TBB library.

Test for persistent_concurrent_hash_map is built and executed only if
Intel TBB library found by CMake.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/libpmemobj-cpp/40)

### CODE REVIEW

File: include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp [hunk 1]
```diff
@@ -0,0 +1,1716 @@
...
+                sz = new_segment.size();
+                if (!new_segment.is_valid()) {
+                    new_segment.enable(pop);
+                }
+
+                init_buckets( pop, new_segment, is_initial );
+                sz <<= 1;// double it to get entire capacity of the container
+            }
+            else { // the first block
+#if INTEL_PRIVATE
```
Comment (pbalcer) [inline include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp]: `#if DEBUG`
do we need this though? I think we should just remove these traces.

File: include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp [hunk 2]
```diff
@@ -0,0 +1,1716 @@
...
+                                            // find next applicable mask after m_old    //TODO: look at bsl instruction
+                for ( ++m_old; !(h & m_old); m_old <<= 1 ) // at maximum few rounds depending on the first block size
+                    ;
+                m_old = (m_old << 1) - 1; // get full mask from a bit
+                __TBB_ASSERT( (m_old&(m_old + 1)) == 0 && m_old <= m, NULL );
+#if INTEL_PRIVATE
+                TRACEP( "%x: mask changed=%x checking with %x\n", unsigned( h ), unsigned( m ), unsigned( m_old ) );
+#endif
+                // check whether it is rehashing/ed
+                if ( itt_load_word_with_acquire( get_bucket( h & m_old )->tmp_node.raw().off ) != rehash_req.raw().off ) // Workaround: just comparing off part. Need to investigate how to properly load with acquire PMEMoid (128 bit)
```
Comment (pbalcer) [inline include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp]: can't you just make tmp_node a persistent pool ptr?

File: include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp [hunk 3]
```diff
@@ -0,0 +1,1716 @@
...
+
+    template<typename Container, typename T, typename U>
+    bool operator!=( const persistent_hash_map_iterator<Container, T>& i, const persistent_hash_map_iterator<Container, U>& j ) {
+        return i.my_node != j.my_node || i.my_map != j.my_map;
+    }
+
+    //! Range class used with persistent_concurrent_hash_map
+    /** @ingroup containers */
+    template<typename Iterator>
+    class hash_map_range {
```
Comment (pbalcer) [inline include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp]: we should probably unify the interfaces with this functionality. in the array it's `slice`

File: include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp [hunk 4]
```diff
@@ -0,0 +1,1716 @@
...
+    result->my_hash = h;
+check_growth:
+    // [opt] grow the container
+    if ( check_growth( m, sz ) ) {
+#if __TBB_STATISTICS
+        my_info_resizes++; // concurrent ones
+#endif
+    }
+
+    // this code is not required since allocated memory automatically deleted if transaction rolled back
```
Comment (pbalcer) [inline include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp]: just remove everything that's unnecessary.

File: include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp [hunk 5]
```diff
@@ -0,0 +1,1716 @@
...
+            p = &n(my_pool_uuid)->next;
+            n = *p;
+        }
+        if ( !n ) { // not found, but mask could be changed
+            if ( check_mask_race( h, m ) )
+                goto restart;
+            return false;
+        }
+        else if ( !b.is_writer() && !b.upgrade_to_writer() ) {
+            if ( check_mask_race( h, m
```
Comment (pbalcer) [inline include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp]: why 512 and 2? make it static const variable.

**Comment by marcinslusarz** [comment]
1) Please make sure all Travis/AppVeyor checks pass.
2) New code should be tested on Travis. This means TBB needs to be installed on our Docker images (see utils/docker/). I'm not sure what to do about AppVeyor...

**Comment by igchor** [comment]
Could you rebase to current master? We have updated mages to ubuntu 18.04 and fedora 28 some time ago.

**Comment by marcinslusarz** [comment]
Please rebase your PR (don't merge master back!) and squash changes into fewer commits.

**Comment by marcinslusarz** [comment]
BTW, you could test your pmemobj_volatile uses which fail right now under pmemcheck with this: https://github.com/pmem/pmdk/pull/3128. I'd like to merge it in 2 days.

**Comment by marcinslusarz** [comment]
*[a discussion](https://reviewable.io/reviews/pmem/libpmemobj-cpp/40#-LJtEAtc1gxAJowkWnG2:-LK2lk8nEv0Qzb14aOeG:b6ubcc9) (no related file):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Stupid question: what is PR stands for?
</blockquote></details>

Pull Request

---

**Comment by marcinslusarz** [comment]
*[a discussion](https://reviewable.io/reviews/pmem/libpmemobj-cpp/40#-LK2kZjx0iD5C-36O5A-:-LK2kZjx0iD5C-36O5A0:bxlisrj) (no related file):*
TBB install from source instructions are confusing, so I installed it from distro package (apt install libtbb-dev).
Result = "Skipping persistent_concurrent_hash_map test because Intel TBB library not  found"

---

**Comment by marcinslusarz** [comment]
Everything related to pmemobj_volatile/v has been merged, so once you rebase your PR, pmemcheck should stop reporting issues related to volatile fields.

**Comment by marcinslusarz** [comment]
___
*[tests/CMakeLists.txt, line 296 at r11](https://reviewable.io/reviews/pmem/libpmemobj-cpp/40#-LL51Ogc9aa7va349h0q:-LLFA6v11ln6XWG0zRsz:blr9xja) ([raw file](https://github.com/pmem/libpmemobj-cpp/blob/fbe6d5fadfa2f4f3c71396397d15663c2b8b99ff/tests/CMakeLists.txt#L296)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please add tests with memcheck, drd and helgrind.
</blockquote></details>

FYI, I fixed a bug in V which your PR triggers under memcheck. See #64.

File: include/libpmemobj++/experimental/concurrent_hash_map.hpp [hunk 6]
```diff
@@ -0,0 +1,3063 @@
...
+#endif
+	std::atomic_thread_fence(std::memory_order_release);
+	location = value;
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+	VALGRIND_HG_ENABLE_CHECKING(&location, sizeof(location));
+#endif
+}
+
+#if _MSC_VER
+static inline unsigned long
```
Comment (lplewa) [inline include/libpmemobj++/experimental/concurrent_hash_map.hpp]: this variant of function returns **unsigned long**, where others returns **int**

File: include/libpmemobj++/experimental/concurrent_hash_map.hpp [hunk 7]
```diff
@@ -0,0 +1,3063 @@
...
+	static inline void
+	__pause(int32_t delay)
+	{
+		for (; delay > 0; --delay) {
+#if _MSC_VER
+			YieldProcessor();
+#elif __GNUC__ && (__i386__ || __x86_64__)
+			// Only i386 and x86-64 have pause instruction
+			__builtin_ia32_pause();
+#endif
```
Comment (lplewa) [inline include/libpmemobj++/experimental/concurrent_hash_map.hpp]: #else
#error "missing pause instruction (...)"
#endif

**Pull request merged**